### ISSUE TITLE
Introduction

### ISSUE THREAD

**Issue by objecttothis** [opensourcepos/opensourcepos#1486]
I apologize if there is a better venue for this, but I wanted to introduce myself.  I am a software developer and the co-owner of a bookstore in Azerbaijan (I am American).  A few years ago I made a few commits, some of which I see were accepted but when I stopped receiving responses from the previous owner of the project on SourceForge I forked the project at 2.0.2 to customize the code to our needs.  I also had very little time to work on the code.  Now my role has changed and I am able to work on the POS project and other code for the bookstore four days per week, so I decided to try to migrate the data to the latest opensourcepos release (successfully) and since it seems like you guys have an established team, I thought I would introduce myself and ask if you have criteria for contributors to the code base.

Since I have multiple projects I'm balancing, the code changes I would want to focus on are the ones that are the union between what benefits our bookstore and the opensourcepos community as a whole.  I realize that some of what may benefit our bookstore may not be in line with the project objectives and therefore would need to be custom and separate from the main code base. I also realize that changes would need to be written in such a way as to be non-specific to the book industry since people from many industries use the software (hence the reason I added the ten custom item fields rather than hard-coded the names of what we use in the bookstore)

Thoughts? questions?

**Comment by daN4cat** [comment]
Collaboration welcome :-)
Custom fields will go soon for proper attributes definition. In the pipeline.

**Comment by objecttothis** [comment]
@daN4cat can you explain what you mean by proper attributes definition?  Also, is there a migration path for the data for users of ospos who rely on those custom fields?

**Comment by jekkos** [comment]
@objecttothis I remember indeed you were frequently on sourceforge. Welcome back! It looks like we'll be having quite some dev capacity for the project for our future time here.

I think in terms of adding new features, it totally makes sense to come and discuss what you are about to build so we can align ideas  on existing functionality and discuss how we can integrate to the community's benefit. 
In the end you will find yourself maybe spending a bit more time (to make things general and reusable),but having code upstreamed is what you get in return (which is other fixes-, contributions beside your own).

So maybe you can enlight us on what you have on the roadmap?
Also custom fields thingy was rewritten by me in a private branch, that's nearly production ready.. First we want to get 3.1 out of the door which is in it's final phase now. Next minor release will probably be that. I foresaw a migration script as we always try to.

**Comment by objecttothis** [comment]
Super, here are the current objectives that I had in no particular order.  I am still getting to know 3.1.0, so some of them were specific to 2.0.2.
- Per-user language setting.  We have several employees with primary languages that are different.  This would keep the language setting in Store-Config->Localisation but allow it to be overridden in the employee config with another language.
- Orders module. One of our major problems is tracking both small special orders and large orders (stock orders and large orders for clients).  In our case this is books, but this really could be used by anyone.  The idea is that when the order is marked as received and checked it gets imported into ospos.  I don't know that this one fits into not making ospos a monolith, so it could potentially be a parallel project that shares data.  Maybe it could be called something like opensourcefrontoffice or something similar.
- Pricing Module.  This one doesn't have a lot of shape in my mind, but is needed.  Right now we have to calculate pricing of our products based on a number of factors (wholesale cost, domestic shipping, international shipping, customs duties, markup, etc).  I'd like to put this module together to help with pricing of inventory.  Since the different types of businesses would have different factors relating to their pricing it would need to be written in a way to have adding of "pricing factors" and defining of those factors (per weight, flat amount, per cost, etc).  This too could go in the parallel frontoffice project that shares data with ospos if it made the project too monolithic.
- Complete translations in Russian and Azerbaijani.
- Customizable "Report Builder".
- GUI OSPOS installer to reduce install problems caused by non-expert (and even expert) users installing but not customizing the install properly.
- Year-end (or any interval) comparison.  Most businesses do some sort of inventory of their stock to correct data and account for theft, breakage.  I was thinking something that accepted a csv file with item_number, name, quantity and retail price and then ran that against the data and spat out potential discrepancies.  This again, could go in the parallel project if did not fit well in OSPOS.
- setup a reward system that offers a discount (percent or amount) after the customer purchase a certain amount (quantity, or cash amount). This would also be customizable to allow purchases to count if conditions are met (such as over a certain price point).  I did see that there is a reward feature currently, so if it doesn't provide these features then perhaps its a matter of simply adding to whats already existing.
- Improve the results window of the item import to show any rows that did not import and show results of what did.
- Allow item import that updates quantities where retail price, name and item_number match. This import feature would either average the wholesale amounts, keep the old wholesale or update with the new. This would either be a separate import from the current import or would be a radio selection between the two types of imports and what to do with the wholesale amounts.  I think beyond that it would fill in blank fields.  Since there is no restriction that any of those three things is unique I think it would only do the import for items where those three were unique and would report which imports were not done.
- Rename cost price to wholesale price.  The term cost price is confusing.  Wholesale price is fairly industry standard.
- allow export of reports to csv/excel
- allow item quantities to be displayed in whole numbers (even if stored as floats) for those business who do not carry inventory that can be a float. (I think maybe this is already implemented?)
- Separate en_GB and en_US translations.  I love my British brethren but I don't know what takings is, and my staff who only speak intermediate level US English definitely don't know what it means.
- Be able to add additional information to suppliers and customers.  Many of our suppliers have multiple phone numbers, different account numbers and different contacts.  It would be nice to have something similar to Android (and I'm sure other OS's) where you can click a + and add a phone number and label for the number or add an email address with label (e.g., accounts receivable department, orders, shipping office, etc)
- I don't know if this is doable (I need to look at the API), but have a checkbox on the edit customer window to "send emails" and have that add them or remove them from the mailchimp list.
- Allow columns to be added and removed in the different views (and also reset).  This is especially useful for users of the custom fields who rely heavily on that data (we need to see author in our item view for example).  I see that its possible to do this, but that custom fields do not appear in the list of columns to add.

There is probably more, but that's what I have for now.

**Comment by objecttothis** [comment]
I know my previous comment is a lot.  Please know that I am very open to adjusting and aligning ideas to best fit what the project needs and not just what my bookstore needs.  If at all possible however, I'm trying to avoid making a custom job because then integrating new versions of ospos becomes a headache, so lets definitely figure out what we can do together... all of the above stuff I am willing to do myself or collaborate.  That is, I'm not expected to dump any of this onto anyone else. :)

**Comment by jekkos** [comment]
Quite a bucket list you have there. I went through it and of course have some thoughts on a possible approach for some. Think we should start with linking these feature with existing issues in our tracker to consolidate design around it

* For inventory we have 
* Adding extra fields to customers , people think should be handled in a generic way. I'd propose to look into the attributes code once it's there. Idea is to make data fields extensible and indeed also configurable in the views and reports, which was already done partly using bootstrap tables #1097, #1193 
* Reports should already be exportable, but only what you currently have loaded in the view (so might be improved) #868, #649 
* Purchase orders has been discussed, I think we need to see what overlaps OSPOS functionality, but some bounded context needs to be taken into account #566, #1479 
* Item CSV result import #1412 
* Stock count functionality (through REST, or CSV) #152
* Report builder using Reportico (@SteveIreland is maybe also working on this?) #1433
* Pricing module #588

**Comment by daN4cat** [comment]
Also we can add discount per customer, adding the rules would be a nice idea.
This point of having a rule engine pops up here and there, it would be good to find some solution (e.g. send SMS to supplier if < restock quantity, send SMS to customer if item in stock, set discount according to x or y)

**Comment by jekkos** [comment]
Most of these rules will probably resolve to querying our domain model (so just generating queries)??
Might be very customizable but also requires a good background on ospos code / design

**Comment by objecttothis** [comment]
@jekkos I do like me my buckets! ;) I'm glad to see so many of those things already on the radar to one degree or another.  Let me get the Azeri, en_US and Russian translations completed and bugs worked out in my data migration then I will be in a much better place to figure out triage things.  I did forget one thing that is a pretty significant item on my roadmap
- CCIdesq (eCommerce CMS) API integration via REST API calls.  Our website is another one of my projects and in order for the inventory for the site to be updated when sales or inventory changes happen on ospos, calls need to be made to the API to update that database so that the two are in sync.  Similar to Mailchimp I planned it to be an optionally enabled API integration.

**Comment by daN4cat** [comment]
Mailchimp integration is done to a certain degree. For instance there is no sync feature if you have already data in OSPOS and data in Mailchimp and those are not aligned as that story gets complicated.
Same as fetching Mailchimp list in OSPOS and the other way round.

**Comment by SteveIreland** [comment]
Seems like 3.2 is going to be an awesome release.  I hope I can keep up!

**Comment by objecttothis** [comment]
Hehe. Out of curiosity, do we have roadmap documented somewhere showing timeline for various milestones?

_Sent from my Android XT1575 using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by daN4cat** [comment]
No documented roadmap and I don't want to start thinking of one. It's up to the point where we have enough to release. Also nobody can commit too much time, so it's a bit pointless to commit for something that you do in your spare time.

**Comment by objecttothis** [comment]
@daN4cat understood.  Let me clarify my statement.  The question about timeline has less to do with meeting deadline and more to do with trying to understand what the collective expectation is for when a milestone is released.  For example if there is a collective expectation that all of my bucket list hits 3.2 in one month I would like to voice that I'm not likely to get that far.  However if the collective expectation is that a community chosen subset of the bucket list is selected for the 3.2 milestone and it was finished when it was finished, then that is much more doable for me.

Separately, my question was unclear but related to the overarching project goals that are in place as something to aim at.  Sort of the adage "if you aim at nothing you hit it every time" comes to mind.  @daN4cat I read a post of yours on #1479 relating to architecture goals.  I think that's what I'm referring to, but I wasn't sure if there was more of that kind of thing documented or discussed somewhere else. I'm also still getting to know the relationship structure and investment of each of the contributors here, so that might be driving my ignorance. What I'm gathering so far (please correct me if I'm wrong) is that you and @jekkos are driving the direction, @SteveIreland has vested interest in the project because he uses the software with clients, but that is about the extent of my knowledge. Sorry, I don't want to bog us down with intros and such when your time is limited and there is much to do.

**Comment by daN4cat** [comment]
@objecttothis no worries and good discussion indeed.
I just wanted to say that we start a release with some sketched ideas and then add stuff in to the point we feel it's enough and mature to release.
The jump to vs 3 was a major drive and we turned around the application. Even at the time nobody thought of a vs 3 like that until I took and said let's do it.
With 3.1 we never planned rewards and somebody came and said I'll do it and contributed.
So with regards to 3.2, we can sketch a couple of high level goals and see if we can make it:
- make taxes more consistent using one approach if @SteveIreland wants to complete that job
- add attributes to items replacing custom fields and stuff like that (for me this is high priority because after two years of work my wife still uses vs 1.1 because I cannot migrate customisation without attributes) @jekkos is working on this
- add Reportico as embedded capability, I started looking into that
- add REST API, but might be 3.3 as this requires some design
- redesign on paper the ideal architecture following a domain model approach and reorganise the data model in the database 
- I know @SteveIreland wants to integrate an Frontaccounting
- would be good to have better permission and access control
- of course anything you want to add from your side
- better migration support and upgrading in the app itself
- maybe a register that has more ajax support as somebody suggested already
- other bits and pieces

Probably we will be able to work on half of this....

Thoughts from the rest of the group?

**Comment by RamkrishnaMondal** [comment]
I don't see conversion of 'Receivings' to 'Purchases' anywhere in the to do list. Got disappointed!

**Comment by SteveIreland** [comment]
With @daN4cat 's assistance I am committing to report builder integration.  I've built a couple of reports with it and although it takes a bit of time to get up to speed with it I'm not disappointed with the results.

I have the CSV import for updating items working for my site, but it uses a vendor id and vendor part number for unique item identification and these are two of the custom fields on the item master.  I'm waiting on the new item attributes feature to be introduced before submitting a pull request.  This change was primarily introduced to update prices where the distributor provides an excel spreadsheet with updated prices.  There is one spreadsheet per vendor and the only reliably unique item identifier is the combination of vendor id and vendor item code.  Your requirements seem to be a bit different and I'd be happy to incorporate your requirements into the change.  I just need a bit more understanding of the business rules involved.

Integrating FrontAccounting is starting to be an unpleasant experience.  I hoped to use an already functioning Rest API to hook in to so that I could avoid spending time learning another PHP framework but I just found out that the  API hasn't been kept current with the latest release.  I'm now trying to decide if it is worth the effort.  Especially if there is going to be an effort made to develop purchasing and receivable modules that are directly based on the OSPOS database.  If that is going to happen I'd much rather be spending my time assisting with the design and development of those efforts since that is the core of what I need from FrontAccounting.

**Comment by daN4cat** [comment]
@RamkrishnaMondal, well, we are awaiting for your contribution in that area :-)

**Comment by daN4cat** [comment]
@SteveIreland if it's easier to add purchasing / receivable in OSPOS, let's do that. In the end it looks like many are asking for such addition. But then more than an POS system it's becoming a retail management system. Maybe that's fine.
I know in my case that I would love to replace a Google spreadsheet with supplier invoices and payment due dates and done with a tool where I can enter invoice data and scan as attachment.
If that becomes part of OSPOS and your suggestion (purchasing) then I can help with that.

BTW, have you added Ciportico to OSPOS already? If yes I won't spend time doing that. I had in mind to do something. Maybe you can submit to a feature/branch so I can give a try. I won't merge that in master until we have release 3.1.

**Comment by RamkrishnaMondal** [comment]
@daN4cat My problem is , I can't implement it myself .... it's beyond my coding abilities.

**Comment by RamkrishnaMondal** [comment]
@SteveIreland @daN4cat  lot of thanks for these lines.... 


> If that is going to happen I'd much rather be spending my time assisting with the design and development of those efforts since that is the core of what I need from FrontAccounting.

> If that becomes part of OSPOS and your suggestion (purchasing) then I can help with that.

**Comment by SteveIreland** [comment]
I added Reportico but backed that out after the deployment architecture change because it had an impact that I didn't understand. after I merged.  So I'll retry this week with Ciportico.  I need to learn how to do that right anyway.

**Comment by daN4cat** [comment]
I can see if I find the time to add Ciportico quickly. I have some Reportico templates that I can easily use for a test.
Also I want to see how the look'n'feel can be made consistent with the OSPOS, I mean the theming part.

**Comment by daN4cat** [comment]
@SteveIreland I can see Reportico guy submitted a rearchitected version in a branch that should support composer. Please put on hold on your side that I will look into this and align everything with the latest from Reportico side.

**Comment by SteveIreland** [comment]
Sounds like a plan.   Thanks.

**Comment by objecttothis** [comment]
Hi everyone. I'm not going to be able to consistently be around for the next week or so. Sadly our 2-year-old son fell out of the 2nd floor window of our friend's house and is in intensive Care at the moment so I need to see to that. I may make a few commits here and there to get my mind off things though. Thank you for understanding.

_Sent from my Android XT1575 using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by daN4cat** [comment]
@objecttothis I'm really sorry to hear that, such a awful incident. I wish your son to get well soon and please stay close to him.

**Comment by daN4cat** [comment]
If the doctors are not against it, read him a book of nice stories. I'm sure he will be happy to hear his daddy reading something to him, and to see you close to him.

Also please don't worry about OSPOS at the moment.

**Comment by objecttothis** [comment]
Good call @daN4cat. I've got my wife bringing a few books to read to him.

_Sent from my Android XT1575 using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by saruban** [comment]
I am praying for your son. Don't worry @objecttothis , he will be alright and play like before happily.

**Comment by objecttothis** [comment]
@saruban Thank you, we appreciate it.  The surgery today went well to repair the bones around his left eye and insert a pin into his left femur.  He will have a bit to heal, but we are so grateful to have not lost him.

**Comment by daN4cat** [comment]
@objecttothis if you don't mind I will remove the personal messages, hope your son is recovering.

**Comment by objecttothis** [comment]
nope, don't mind.  He is doing well.  Thank you.

**Comment by daN4cat** [comment]
@objecttothis I took the initiative to change your list to a task list with checkboxes, so we can tick out your list when something is completed. Maybe you want to recheck and see if correct or now.

**Comment by objecttothis** [comment]
Good call.  I'll update it.

**Comment by daN4cat** [comment]
Old thread, looks like it's time to close it.
